### DISCUSSION TITLE
Helm uninstalling EMQX Operator deletes the EMQX instances

### DISCUSSION THREAD

**Discussion by ennc0d3** [emqx/emqx-operator#1059] [General]
I have installed v2.2.23 version of the emqx-operator and deployed the EMQX by creating the EMQX CRD instances (Core/Replica) node.

The EMQX instances are running successfully, but I am expecting that the deletion of helm release (i.e the emqx-operator) should not delete instances managed by the operator it should only remove EMQX Operator. Unfortunately it is removing all EMQX instances, why is it so and is it possible to change the behavior to delete only the operator?

 
Thanks, Naga

**Comment by Rory-Z** [comment]
Hi @ennc0d3 Delete Helm chart will be delete all of CRD，and then，k8s will be delete all of CR.
I don't understand, why are you need delete operator controller and keep CR ? because if operator controller is not running, the EMQX CR perhaps have wrong status.

**Reply by ennc0d3** [reply]
The reason is let’s say I want to migrate the operator from one namespace to another. I want to delete the operator running in NS1 without affecting the EMQX CR and the resources created due to it. Then install the operator in another namespace NS2 which will take over the CR.

Are there any labels that I can use so that the operator doesn’t kill the EMQX? Any other ideas?

Thanks,

**Reply by Rory-Z** [reply]
Maybe you can install CRD first, you can got them by `helm template`, and second, you can install EMQX operator and set `--skipCRDs=true`.
I think it will be work.

**Reply by ennc0d3** [reply]
I got it working by following these steps:

**Edit the CRDs in the emqx-operator chart:** Add the annotation` helm.sh/resource-policy: keep` to the CRDs. This edit is necessary because Helm ignores annotations that are manually applied using kubectl.

**Upgrade the chart:** This step applies the annotation to the CRDs. Once annotated, Helm will retain these CRDs even if the release is uninstalled.

**Install the chart in a new namespace with skipCRDs=true:** Before installation, annotate a few resources (clusterrole, clusterolebinding, mutatingwebhookconfiguration, validatingwebhookconfigurations) with `meta.helm.sh/release-namespace=newNS`.

**Verify the CRD annotations:** Ensure that the `helm.sh/resource-policy: keep ` annotation remains intact.

The key aspect is the [resource-policy annotation](https://helm.sh/docs/howto/charts_tips_and_tricks/#tell-helm-not-to-uninstall-a-resource). IMHO, it would be beneficial to have a flag to control the deletion of CRDs. When this flag is set, we can use the annotation to retain the CRDs.

Thanks, Naga
